core-api/src/main/java/org/silverpeas/core/persistence/jdbc/sql/DefaultJdbcSqlExecutor.java

```diff
@@ -73,38 +81,55 @@ public long selectCount(final Connection con, final JdbcSqlQuery selectCountQuer
           throw new IllegalArgumentException("select count execution error");
         }
         return count;
+      } catch (SQLException e) {
+        SilverLogger.getLogger(this)
+            .debug(e.getMessage() + SQL_REQUEST + selectCountQueryBuilder.getSqlQuery());
+        throw e;
       }
     }
   }
 
   @Override
-  public <R> List<R> select(JdbcSqlQuery selectQuery, SelectResultRowProcess<R> process)
+  public <R> ListSlice<R> select(JdbcSqlQuery selectQuery, SelectResultRowProcess<R> process)
       throws SQLException {
     try (Connection con = ConnectionPool.getConnection()) {
       return select(con, selectQuery, process);
     }
   }
 
   @Override
-  public <R> List<R> select(final Connection con, final JdbcSqlQuery selectQuery,
+  public <R> ListSlice<R> select(final Connection con, final JdbcSqlQuery selectQuery,
       final SelectResultRowProcess<R> process) throws SQLException {
-    try (PreparedStatement st = con.prepareStatement(selectQuery.getSqlQuery())) {
+    JdbcSqlQuery.Configuration queryConf = selectQuery.getConfiguration();
+    try (PreparedStatement st = queryConf.isFirstResultScrolled() ?
+        con.prepareStatement(selectQuery.getSqlQuery(), ResultSet.TYPE_SCROLL_INSENSITIVE,
+            ResultSet.CONCUR_READ_ONLY) : con.prepareStatement(selectQuery.getSqlQuery())) {
       setParameters(st, selectQuery.getParameters());
       try (ResultSet rs = st.executeQuery()) {
-        List<R> entities = new ArrayList<>();
-        int i = 0;
-        while (rs.next()) {
-          int resultLimit = selectQuery.getConfiguration().getResultLimit();
-          if (resultLimit > 0 && entities.size() >= resultLimit) {
-            break;
-          }
-          R entity = process.currentRow(new ResultSetWrapper(rs, i));
-          if (entity != null) {
-            entities.add(entity);
+        int idx = queryConf.getOffset();
+        if (queryConf.isFirstResultScrolled()) {
+          rs.next();
+          rs.relative(idx - 1);
+        }
+        final int lastIdx =
+            queryConf.isResultCountLimited() ? idx + queryConf.getResultLimit() - 1 : 0;
+        ListSlice<R> entities = new ListSlice<>(idx, lastIdx);
+        int originalSize = idx;
+        for (;rs.next(); originalSize++) {
+          if (!queryConf.isResultCountLimited() ||
+              (queryConf.isResultCountLimited() && entities.size() < queryConf.getResultLimit())) {
+            R entity = process.currentRow(new ResultSetWrapper(rs, idx++));
+            if (entity != null) {
```

Comment:
![MINOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png 'Severity: MINOR') Refactor this code to not nest more than 4 if/for/while/switch/try statements. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://integration.silverpeas.org/sonar/coding_rules#rule_key=squid%3AS134)